code/framework/java-server/src/main/java/io/github/ibuildthecloud/gdapi/request/parser/DefaultApiRequestParser.java

```diff
@@ -160,22 +162,76 @@ protected String parseClientIp(ApiRequest apiRequest, HttpServletRequest request
         String clientIp = request.getRemoteAddr();
 
         clientIp = getOverrideHeader(request, overrideClientIpHeader, clientIp);
-        clientIp = getOverrideHeader(request, FORWARDED_HEADER, clientIp, false);
+        clientIp = getOverrideHeader(request, FORWARDED_FOR_HEADER, clientIp, false);
 
         return clientIp;
     }
 
+    /**
+     * Constructs the request URL based off of standard headers in the request, falling back to the HttpServletRequest.getRequestURL() 
+     * if the headers aren't available. Here is the ordered list of how we'll attempt to construct the URL: 
+     *  - x-api-request-url
+     *  - x-forwarded-proto://x-forwarded-host:x-forwarded-port/HttpServletRequest.getRequestURI()
+     *  - x-forwarded-proto://x-forwarded-host/HttpServletRequest.getRequestURI() 
+     *  - x-forwarded-proto://host:x-forwarded-port/HttpServletRequest.getRequestURI()
+     *  - x-forwarded-proto://host/HttpServletRequest.getRequestURI() request.getRequestURL()
+     * 
+     * Additional notes: 
+     *  - With x-api-request-url, the query string is passed, it will be dropped to match the other formats. 
+     *  - If the x-forwarded-host/host header has a port and x-forwarded-port has been passed, x-forwarded-port will be used.
+     */
     protected String parseRequestUrl(ApiRequest apiRequest, HttpServletRequest request) {
-        String requestUrl = null;
-        try {
-            requestUrl = URLDecoder.decode(request.getRequestURL().toString(), "UTF-8");
-        } catch (UnsupportedEncodingException e) {
-            throw new IllegalStateException(e);
+        // Get url from custom x-api-request-url header
+        String requestUrl = getOverrideHeader(request, overrideUrlHeader, null);
+        if (requestUrl != null) {
+            String[] parts = requestUrl.split("\\?", 2);
+            return parts[0];
+        }
+
+        // Get url from standard headers
+        requestUrl = getUrlFromStandardHeaders(request);
+        if (requestUrl != null) {
+            return requestUrl;
+        }
+
+        // Use incoming url
+        return request.getRequestURL().toString();
+    }
+
+    private String getUrlFromStandardHeaders(HttpServletRequest request) {
+        String xForwardedProto = getOverrideHeader(request, FORWARDED_PROTO_HEADER, null, false);
+        if (xForwardedProto == null) {
+            return null;
+        }
+
+        String host = getOverrideHeader(request, FORWARDED_HOST_HEADER, null, false);
+        if (host == null) {
+            host = getOverrideHeader(request, HOST_HEADER, null, false);
         }
 
-        requestUrl = getOverrideHeader(request, overrideUrlHeader, requestUrl);
+        if (host == null) {
```

Comment:
Yeah, I went back and forth on that. Decided against it because it seemed very unlikely it would be empty and I was trying to be as minimalistic as possible in my checks since this is ran on every api request.

Thank it is better to just be safe and make it a StringUtils.isEmpty() check?